src/Covenant/JSON.hs

```diff
@@ -0,0 +1,517 @@
+    _asg :: Map Id ASGNode,
+    _version :: Version
+  }
+  deriving stock (Show, Eq)
+
+{- CompilationUnit
+
+   Encodes as an object. The maps are represented by KV pairs in arrays. Example:
+
+   {datatypes: [{k: "Maybe", v: ...}, {k: "Foo", v: ...}],
+    asg: [{k: 0, v: ...}],
+    version: {major: 1, minor: 2}
+   }
+-}
+
+instance ToJSON CompilationUnit where
+  toEncoding = error "TODO: Implement toEncoding for CompilationUnit"
+  toJSON = error "TODO: Implement toJSON for CompilationUnit"
+
+instance FromJSON CompilationUnit where
+  parseJSON = error "TODO: Implement fromEncoding for CompilationUnit"
+
+-- TODO: Remove this after I'm certain I don't need to write any more of these -_-
+mkStub :: String -> String
+mkStub ty = commentsBlock <> encodingStub <> "\n\n" <> decodingStub
+  where
+    commentsBlock = "{- " <> cleanName <> "\n\n-}\n"
+    cleanName = filter (/= ' ') ty
+    encFn = "encode" <> cleanName
+    decFn = "decode" <> cleanName
+    encodingStub =
+      encFn
+        <> " :: "
+        <> ty
+        <> " -> "
+        <> "Encoding"
+        <> "\n"
+        <> encFn
+        <> " = error \"TODO: Implement "
+        <> encFn
+        <> "\""
+    decodingStub =
+      decFn
+        <> " :: Value -> ASGParser ("
+        <> ty
+        <> ")"
+        <> "\n"
+        <> decFn
+        <> " = error \"TODO: Implement "
+        <> decFn
+        <> "\""
+
+{- Version
+
+   Serializes as an object with the fields you'd expect.
+
+     Version 1 2
+   ->
+     {major: 1, minor: 2}
+
+-}
+encodeVersion :: Version -> Encoding
+encodeVersion = error "TODO: Implement encodeVersion"
+
+decodeVersion :: Value -> ASGParser Version
+decodeVersion = error "TODO: Implement decodeVersion"
+
+{- DataDeclaration & its components -}
+
+{- Encodes as a simple JSON string, e.g.
+   TyName "Foo" -> "Foo"
+-}
+encodeTyName :: TyName -> Encoding
+encodeTyName = error "TODO: Implement encodeTyName"
+
+-- NOTE: Need to validate that the string conforms w/ the Type naming rules
+decodeTyName :: Value -> ASGParser TyName
+decodeTyName = error "TODO: Implement decodeTyName"
+
+{- Encodes as a simple JSON string, e.g.
+   ConstructorName "Foo" -> "Foo"
+-}
+encodeConstructorName :: TyName -> Encoding
+encodeConstructorName = error "TODO: Implement encodeConstructorName"
+
+-- NOTE: Need to validate that the string conforms w/ the Type naming rules
+decodeConstructorName :: Value -> ASGParser ConstructorName
+decodeConstructorName = error "TODO: Implement decodeConstructorName"
+
+{- Encodes as an object. E.g.:
+
+   Constructor "Just" [IntegerT]
+   ->
+   { constructorName: "Just"
+   , constructorArgs: [...]}
+
+-}
+encodeConstructor :: Constructor AbstractTy -> Encoding
+encodeConstructor = error "TODO: Implement encodeConstructor"
+
+decodeConstructor :: Value -> ASGParser (Constructor AbstractTy)
+decodeConstructor = error "TODO: Implement decodeConstructor"
+
+{- DataEncoding encodes as a typical sum type, and will look like:
```

Comment:
We don't specify what a 'typical' sum type encoding is. Probably needs a module-level doc comment.